buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersRegistry.java

```diff
@@ -0,0 +1,67 @@
+package org.elasticsearch.gradle.testclusters;
+
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class TestClustersRegistry {
+    private static final Logger logger =  Logging.getLogger(TestClustersRegistry.class);
+    private static final String TESTCLUSTERS_INSPECT_FAILURE = "testclusters.inspect.failure";
+    private final Boolean allowClusterToSurvive = Boolean.valueOf(System.getProperty(TESTCLUSTERS_INSPECT_FAILURE, "false"));
+    private final Map<ElasticsearchCluster, Integer> claimsInventory = new HashMap<>();
+    private final Set<ElasticsearchCluster> runningClusters = new HashSet<>();
+
+    public void claimCluster(ElasticsearchCluster cluster) {
+        cluster.freeze();
+        claimsInventory.put(cluster, claimsInventory.getOrDefault(cluster, 0) + 1);
+    }
+
+    public void maybeStartCluster(ElasticsearchCluster cluster) {
+        // we only start the cluster before the actions, so we'll not start it if the task is up-to-date
+        if (runningClusters.contains(cluster)) {
+            return;
+        }
+        runningClusters.add(cluster);
+        cluster.start();
+    }
+
+    public void stopCluster(ElasticsearchCluster cluster, boolean taskFailed) {
+        if (taskFailed) {
```

Comment:
This condition is backwards, it just so happens to work because we are also inverting the `taskFailed` predicate at the call site.